src/librustc_mir/borrow_check.rs

```diff
@@ -886,17 +903,43 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
                     err.emit();
                 }
             },
-            _ => {}// Access authorized
+            Write(WriteKind::Mutate) => {
+                if let Err(lvalue_err) = self.is_mutable(lvalue, is_local_mutation_allowed) {
+                    let mut err = self.tcx.cannot_assign(span,
+                        &format!("immutable item `{}`",
+                                self.describe_lvalue(lvalue)),
+                        Origin::Mir);
+                    err.span_label(span, "cannot mutate");
+
+                    if lvalue != lvalue_err {
+                        err.note(&format!("Value not mutable causing this error: `{}`",
+                            self.describe_lvalue(lvalue_err)));
+                    }
+
+                    err.emit();
+                }
+            },
+            Write(WriteKind::MutableBorrow(BorrowKind::Shared)) |
```

Comment:
This should be a `delay_span_bug` if this is not an immutable local in "no immutable local" mode - all 3 of these cases indicate an erroneous program.